### ISSUE TITLE
Default to DELETE instead of TRUNCATE for Sweep

### ISSUE THREAD

**Issue by pwilczynski** [palantir/atlasdb#1351]
While truncate is definitely a faster option for sweep / scrub, it might be nice to have AtlasDB default to using DELETE. This will be slower, but guaranteed not to break backups. This seems more optimal than having to turn off Sweep/Scrub during backups, and teams that need faster sweep / scrub could use TRUNCATE at the cost of needing to take out a Backup Lock before running the dump, and remembering to release said lock after the backup job completed.

**Comment by rhero** [comment]
Just adding that I think this is only applicable for Oracle and Postgres, and this isn't possible for Cassandra.

**Comment by tkleingarn** [comment]
Now closed, an initial PR from @carrino is https://github.com/palantir/atlasdb/pull/1321.

Tagging @clockfort for SA who is planning to pick up work on this.  @rhero ACK, this applies to Oracle and Postgres.

**Comment by hsaraogi** [comment]
Blocked on #1391.

**Comment by gsheasby** [comment]
Not blocked anymore! @clockfort, are you going to pick this up?

**Comment by clockfort** [comment]
I would prefer to impl this as "if user calls truncate, try to truncate, but if the call fails, retry as a RangeRequest.deleteRange(all) and issue a log at WARN", especially given how bad people are at sweeping large amounts of data.

Does that work for people?

**Comment by gsheasby** [comment]
`kvs.deleteRange(table, RangeRequest.all())` is effectively the same as truncating, right? Why is it more likely to succeed?

Does the suggestion work for @pwilczynski?

**Comment by clockfort** [comment]
> kvs.deleteRange(table, RangeRequest.all()) is effectively the same as truncating, right? Why is it more likely to succeed?

Only from the view of the DB client; in the DB itself it's usually the difference between 
- a DDL operation that needs (Oracle/Postgres: Table Lock, DDL change that breaks some backups, C* a cluster-wide schema mutation and agreement round) 
and
- a delete operation that (Oracle/Postgres: Builds up a lot of undo / stuff to vacuum, C*: tombstones) and will take longer to perform (linear to the size of the table)

**Comment by clockfort** [comment]
After reviewing, my impl won't work for Peter's use case though (not breaking backups) since the backups breakage doesn't boil up any error to Atlas, only to the backup. :(

New idea: Can we make sweep attempt to get the backup lock, and if it gets it, do a truncate, and if it doesn't get it, does delete?

**Comment by gsheasby** [comment]
Why can't sweep simply wait for the backup to finish before carrying on and going the truncate? That's the current functionality.

**Comment by tpetracca** [comment]
@clockfort that means we'd have to make the sweep code aware of what impl of the KVS you were using, as we still don't want to issue any deletes at all if you don't hold the lock in the case of CassandraKVS

**Comment by tpetracca** [comment]
I think the answer here is simple that we always does one or the other (range delete vs truncate) on the basis of some configuration endpoint that is DbKvs specific, or our large internal product simply uses the backup lock....

**Comment by pwilczynski** [comment]
Yeah, if we had an endpoint, then we could flip it manually as part of taking backups.

**Comment by tpetracca** [comment]
@pwilczynski it's merged! https://palantir.github.io/atlasdb/html/cluster_management/backup-restore.html#obtain-the-backup-lock

**Comment by rhero** [comment]
It seems like the only reason we wouldn't want Cassandra to use deleteRange for Sweep is because of performance, but those concerns seem fairly small, correct? 

Sounds like if we still grab the backup lock before sweeping, and always perform deleteRange, then we will satisfy all cases (Cassandra, Postgres, and Oracle) where Sweep interferes with backups.

**Comment by berler** [comment]
Re: @gsheasby on "Why can't sweep simply wait for the backup to finish before carrying on and going the truncate? That's the current functionality."

Most of our users perform their own backups, and they typically do this using standard tools.  In many cases the people setting up or performing DB backups are DBAs that only have access directly to the DB and do not ever do things with our applications (they might not even have access so they will never run any sort of CLI that takes out the backup lock).  In general we cannot make the assumption that only our own applications are going to talk to the DB, or that people will only use our "supported" backup method.

**Comment by jboreiko** [comment]
Just to close this out.  Sweep currently acquires the backup lock prior to running the actual kvs#delete()s and releases after.  Outside of this kvs#truncate is never run by either the SweepCommand or the BackgroundSweeperImpl.  Instead kvs#deleteRange is used to clear the sweep.progress table (this was the cause of the db lock contention with backups).  As a note, kvs#deleteRange in Cassandra does a truncate if the range is all (for performance reasons), the postgres and oracle impls do not use this optimization to avoid the original contention.
